File: 2015/06/solution.rb
class LightGrid

  INSTRUCTION_MATCHER = /^(.*) (\d+),(\d+) through (\d+),(\d+)$/

  def initialize(toggle_on_adjustment, toggle_off_adjustment, max_brightness = nil)
    @toggle_on_adjustment = toggle_on_adjustment
    @toggle_off_adjustment = toggle_off_adjustment
    @max_brightness = max_brightness
    @lights = Array.new(1000) do
      Array.new(1000, 0)
    end
  end

  def solve
    File.open('input.txt').each_line do |line|
      instr = INSTRUCTION_MATCHER.match(line)[1..5]

      fn = case instr.shift
        when 'turn on'
          :turn_on
        when 'turn off'
          :turn_off
        when 'toggle'
          :toggle
      end
      
      x1, y1, x2, y2 = instr.map(&:to_i)

      for x in x1..x2
        for y in y1..y2
          send(fn, x, y)
        end
      end
    end

    sum_brightness
  end

  protected
  def sum_brightness
    total = 0
    @lights.each do |row|
      (row || []).each do |light|
        total += light ? light : 0
      end
    end
    total
  end

  def adjust(x, y, brightness)
    value = [ 0, @lights[x][y] + brightness ].max
    value = [ value, @max_brightness ].min if @max_brightness
    @lights[x][y] = value
  end

  def turn_on(x, y)
    adjust(x, y, 1)
  end

  def turn_off(x, y)
    adjust(x, y, -1)
  end

  def toggle(x, y)
    adjust(x, y, @lights[x][y] > 0 ? @toggle_off_adjustment : @toggle_on_adjustment)
  end

end

puts 'Part 1', LightGrid.new(1, -1, 1).solve
puts 'Part 2', LightGrid.new(2, 2).solve

File: 2015/06/solution.js
const solve = (toggleOnAdjustment, toggleOffAdjustment, maxBrightness = undefined) => {

  const INSTRUCTION_MATCHER = /^(.*) (\d+),(\d+) through (\d+),(\d+)$/

  let lights = []

  const sumBrightness = () => {
    let total = 0
    lights.forEach((row) => {
      row.forEach((light) => { total += light ? light : 0 })
    })
    return total
  }

  const brightness = (x, y) => {
    const value = (lights[x] || [])[y]
    return value ? value : 0
  }

  const adjust = (x, y, brightness) => {
    if (!lights[x]) { lights[x] = [] }
    if (!lights[x][y]) { lights[x][y] = 0 }
    let value = Math.max(0, lights[x][y] + brightness)
    if (maxBrightness) {
      value = Math.min(value, maxBrightness)
    }
    lights[x][y] = value
  }

  const turnOn = (x, y) => adjust(x, y, 1)

  const turnOff = (x, y) => adjust(x, y, -1)

  const toggle = (x, y) => {
    const lightOn = brightness(x, y) > 0
    adjust(x, y, lightOn ? toggleOffAdjustment : toggleOnAdjustment)
  }

  const dispatchInstruction = (line) => {
    let [, instruction, x1, y1, x2, y2 ] = line.match(INSTRUCTION_MATCHER) || []
    let instructionFn = ((instruction) => {
      switch (instruction) {
        case 'turn on':
           return turnOn
        case 'turn off':
          return turnOff
        case 'toggle':
          return toggle
        default:
          throw `Unexpected instruction: ${instruction} [input = "${line}"]`
      }
    })(instruction)

    x1 = parseInt(x1, 10)
    x2 = parseInt(x2, 10)
    y1 = parseInt(y1, 10)
    y2 = parseInt(y2, 10)
    
    for (let x = x1; x <= x2; x++) {
      for (let y = y1; y <= y2; y++) {
        instructionFn(x, y)
      }
    }
  }

  const report = () => { console.log(`Total brightness is ${sumBrightness()}`) }

  require('readline').
      createInterface({ input: require('fs').createReadStream('input.txt') }).
      on('line', dispatchInstruction).
      on('close', report)
}

solve(1, -1, 1)
solve(2, 2)
